File: A1/37peg/src/board.rs
use std::vec::Vec;

pub struct Board {
    pub id: u64,
    pub score: u8
}

impl Board {
    pub fn new(state:u64) -> Board {
        let score = Board::count_zeros(state);
        return Board {id: state, score: score};
    }

    pub fn id_to_array(id:u64) -> [[u8;7];7] {
        let mut state:[[u8;7];7] = [[0;7];7];
        let mut bit:u64 = 1 << 48;

        for row in 0..state.len() {
            for col in 0..state.len() {
                match id&bit {
                    0 => state[row][col] = 0,
                    _ => state[row][col] = 1
                }

                bit >>= 1;
            }
        }

        //Top row
        state[0][0] = 9;
        state[0][1] = 9;
        state[0][5] = 9;
        state[0][6] = 9;

        //Second row
        state[1][0] = 9;
        state[1][6] = 9;

        //Second last row
        state[5][0] = 9;
        state[5][6] = 9;

        //Bottom row
        state[6][0] = 9;
        state[6][1] = 9;
        state[6][5] = 9;
        state[6][6] = 9;

        return state;
    }

    pub fn array_to_id(array:&[[u8;7]; 7]) -> u64 {
        let mut id:u64 = 0;

        for row in 0..array.len() {
            for col in 0..array.len() {
                match array[row][col] {
                    1 => id = (id|1) << 1,
                    _ => id <<= 1
                }
            }
        }
        id >>= 1;
        
        return id;
    }

    pub fn count_zeros(id: u64) -> u8 {
        let mut bit:u64 = 1 << 48;
        let mut zeros:u8 = 0;

        for _row in 0..7 {
            for _col in 0..7 {
                let val = id&bit;

                if val == 0 { zeros += 1; }

                bit >>= 1
            }
        }

        zeros-=12;

        return zeros;
    }

    pub fn get_children(&self) -> Vec<u64> {
        let state: u64 = self.id;
        let mut arr = Board::id_to_array(state);
        let mut children:Vec<u64> = Vec::new();

        for row in 0..arr.len() {
            for col in 0..arr.len() {
                //Peg or invalid
                if arr[row][col] == 9 || arr[row][col] == 1 { continue; }
                else if arr[row][col] == 0 {
                    
                    //Check Up - Jump Down
                    if row > 1 && arr[row-1][col] == 1 && arr[row-2][col] == 1 {
                        arr[row][col] = 1;
                        arr[row-1][col] = 0;
                        arr[row-2][col] = 0;

                        children.push(Board::array_to_id(&arr));

                        arr[row][col] = 0;
                        arr[row-1][col] = 1;
                        arr[row-2][col] = 1;
                    }

                    //Check Right - Jump Left
                    if col < 5 && arr[row][col+1] == 1 && arr[row][col+2] == 1 {
                        arr[row][col] = 1;
                        arr[row][col+1] = 0;
                        arr[row][col+2] = 0;

                        children.push(Board::array_to_id(&arr));

                        arr[row][col] = 0;
                        arr[row][col+1] = 1;
                        arr[row][col+2] = 1;
                    }

                    //Check Down - Jump Up
                    if row < 5 && arr[row+1][col] == 1 && arr[row+2][col] == 1{
                        arr[row][col] = 1;
                        arr[row+1][col] = 0;
                        arr[row+2][col] = 0;

                        children.push(Board::array_to_id(&arr));

                        arr[row][col] = 0;
                        arr[row+1][col] = 1;
                        arr[row+2][col] = 1;
                    }

                    //Check Left - Jump Right
                    if col > 1 && arr[row][col-1] == 1 && arr[row][col-2] == 1{
                        arr[row][col] = 1;
                        arr[row][col-1] = 0;
                        arr[row][col-2] = 0;

                        children.push(Board::array_to_id(&arr));

                        arr[row][col] = 0;
                        arr[row][col-1] = 1;
                        arr[row][col-2] = 1;
                    }
                }
            }
        }

        return children;
    }
}

File: A2/js/board.js
function StartState() {
    return [
        [1, 1, 2, 1, 1, 2],
        [2, 2, 1, 1, 1, 2],
        [2, 2, 1, 2, 2, 1],
        [1, 2, 2, 1, 2, 2],
        [2, 1, 1, 1, 2, 2],
        [2, 1, 1, 2, 1, 1]
    ];
}


class Board {
    constructor(state, turn, parent) {
        this.state = copy2DArray(state);
        this.turn = turn;
        this.parent = parent;
        this.children = [];
        this.generated = false;
    }

    setScore() {
        this.score = this.getScore();
    }

    getScore(heuristic) {
        let score = 0;
        const state = this.state;

        if (this.isWinState()) return Number.NEGATIVE_INFINITY;
        if (this.isLoseState()) return Number.NEGATIVE_INFINITY;

        if (heuristic === 1 || heuristic === 3) {
            for (var row = 0; row < state.length; row++) {
                for (var col = 0; col < state[row].length; col++) {
                    if (state[row][col] == this.turn) {
                        score++
                    }
                    else if (state[row][col] !== 0) {
                        score--;
                    }
                }
            }
        }

        if (heuristic === 2 || heuristic === 3) {
            let temp = 0;

            for (var p of this.getPieces()) {
                const { row, col } = p;

                //all the way left
                for (var i = col; i >= 0; i--) {
                    if (state[row][i] === this.turn) temp--;
                    if (state[row][i] !== 0) temp++;
                }


                //all the way right
                for (var i = col; i < 6; i++) {
                    if (state[row][i] === this.turn) temp--;
                    if (state[row][i] !== 0) temp++;
                }

                //all the way up
                for (var i = row; i >= 0; i--) {
                    if (state[i][col] === this.turn) temp--;
                    if (state[i][col] !== 0) temp++;
                }

                //all the way down
                for (var i = 0; i < 6; i++) {
                    if (state[i][col] === this.turn) temp--;
                    if (state[i][col] !== 0) temp++;
                }

                if (heuristic === 3) temp = temp / 3;

                score += temp;
            }
        }

        return score;
    }

    getRandomChild() {
        const state = this.state;

        this.populateChildren();
        return this.children[randomNum(this.children.length)];

        function randomNum(hi) {
            return Math.floor((Math.random() * hi));
        }
    }

    getChildren() {
        this.populateChildren();
        return this.children;
    }

    populateChildren() {
        if (this.generated) { return; }

        if (this.isWinState() || this.isLoseState()) {
            return;
        }

        this.generated = true;

        const state = this.state;
        let pieces = this.getPieces();
        let children = [];

        for (var p of pieces) {
            let check, mx, length, pad;
            const { row, col } = p;

            //Upwards slide
            check = getColumn(state, col);
            mx = copyArray(check);
            mx = mx.slice(0, row + 1);
            length = mx.length;

            mx = mx.filter(el => el !== 0);
            length = length - mx.length;

            pad = [];
            for (var i = 0; i < length; i++) pad.push(0);
            pad = pad.concat(check.slice(row + 1));

            if (mx[0] !== this.turn && mx[0] !== 0) {
                while (mx[0] !== this.turn) {
                    mx.shift();
                    mx.push(0);
                    let temp = copy2DArray(this.state)
                    temp = replaceColumn(temp, mx.concat(pad), col);
                    children.push(temp);
                }
            }

            // //Downwards slide
            check = copyArray(getColumn(copy2DArray(state), col));
            mx = copyArray(check);
            mx = mx.slice(row);
            length = mx.length;

            mx = mx.filter(el => el !== 0);
            length = length - mx.length;

            pad = [];
            for (var i = 0; i < length; i++) pad.push(0);
            pad = check.slice(0, row).concat(pad);

            if (mx[mx.length - 1] !== this.turn && mx[mx.length - 1] !== 0) {
                while (mx[mx.length - 1] !== this.turn) {
                    mx.unshift(0);
                    mx.pop();
                    let temp = copy2DArray(this.state)
                    temp = replaceColumn(temp, pad.concat(mx), col);
                    children.push(temp);
                }
            }

            //Left slide
            check = copyArray(state[row]);
            mx = copyArray(check);
            mx = mx.slice(0, col + 1);
            length = mx.length;

            mx = mx.filter(el => el !== 0);
            length = length - mx.length;

            pad = [];
            for (var i = 0; i < length; i++) pad.push(0);
            pad = pad.concat(check.slice(col + 1));

            if (mx[0] !== this.turn && mx[0] !== 0) {
                while (mx[0] !== this.turn) {
                    mx.shift();
                    mx.push(0);
                    let temp = copy2DArray(this.state)
                    temp[row] = mx.concat(pad);
                    children.push(temp);
                }
            }

            //Right slide
            check = copyArray(state[row]);
            mx = copyArray(check);
            mx = mx.slice(col);
            length = mx.length;

            mx = mx.filter(el => el !== 0);
            length = length - mx.length;

            pad = [];
            for (var i = 0; i < length; i++) pad.push(0);
            pad = check.slice(0, col).concat(pad);

            if (mx[mx.length - 1] !== this.turn && mx[mx.length - 1] !== 0) {
                while (mx[mx.length - 1] !== this.turn) {
                    mx.unshift(0);
                    mx.pop();
                    let temp = copy2DArray(this.state)
                    temp[row] = pad.concat(mx);
                    children.push(temp);
                }
            }

            //Single moves
            if (row != 5 && state[row + 1][col] == 0) {
                let temp = copy2DArray(state)
                temp[row + 1][col] = this.turn;
                temp[row][col] = 0;
            }

            if (row != 0 && state[row - 1][col] == 0) {
                let temp = copy2DArray(state)
                temp[row - 1][col] = this.turn;
                temp[row][col] = 0
                children.push(temp)
            }

            if (col != 5 && state[row][col + 1] == 0) {
                let temp = copy2DArray(state)
                temp[row][col + 1] = this.turn;
                temp[row][col] = 0
                children.push(temp)
            }

            if (col != 0 && state[row][col - 1] == 0) {
                let temp = copy2DArray(state)
                temp[row][col - 1] = this.turn;
                temp[row][col] = 0
                children.push(temp)
            }
        }

        for (var c of children) {
            const turn = (this.turn === 1) ? 2 : 1;
            this.children.push(new Board(c, turn, this));
        }
    }

    getPieces() {
        let pieces = [];
        const state = this.state;

        for (var row = 0; row < state.length; row++) {
            for (var col = 0; col < state[row].length; col++) {
                if (state[row][col] == this.turn) {
                    pieces.push({ row, col });
                }
            }
        }

        return pieces;
    }

    isWinState() {
        const state = this.state;
        const enemy = (this.turn === 1) ? 2 : 1;

        for (var row = 0; row < state.length; row++) {
            for (var col = 0; col < state[row].length; col++) {
                if (state[row][col] !== enemy) {
                    return false;
                }
            }
        }
        return true;
    }

    isLoseState() {
        const state = this.state;

        for (var row = 0; row < state.length; row++) {
            for (var col = 0; col < state[row].length; col++) {
                if (state[row][col] === this.turn) {
                    return false;
                }
            }
        }
        return true;
    }
}

function getColumn(arr, index) {
    return arr.map((val, ind) => { return val[index]; });
}

function replaceColumn(arr, newCol, index) {
    return arr.map((val, ind) => {
        val[index] = newCol[ind];
        return val;
    });
}

function copyArray(array) {
    return array.slice(0);
}

function copy2DArray(array) {
    return array.map(function (arr) {
        return arr.slice(0);
    });
}

function checkUndefined(state) {
    for (var row = 0; row < state.length; row++) {
        for (var col = 0; col < state[row].length; col++) {
            if (state[row][col] == undefined) {
                return true;
            }
        }
    }
    return false;
}

module.exports = {
    StartState,
    Board
}